tests/functional/src/main/java/org/mule/functional/junit4/ExtensionsTestInfrastructureDiscoverer.java

```diff
@@ -0,0 +1,222 @@
+ * <p/>
+ * Once extensions are registered, a {@link ResourcesGenerator} is used to automatically
+ * generate any backing resources needed (XSD schemas, spring bundles, etc).
+ * <p/>
+ * In this way, the user experience is greatly simplified when running the test
+ * either through an IDE or build tool such as maven or gradle.
+ * <p/>
+ *
+ * @since 4.0
+ */
+public class ExtensionsTestInfrastructureDiscoverer
+{
+    private final ServiceRegistry serviceRegistry = new SpiServiceRegistry();
+    private final ExtensionFactory extensionFactory = new DefaultExtensionFactory(serviceRegistry, getClass().getClassLoader());
+    private final ExtensionManagerAdapter extensionManager;
+    private final File generatedResourcesDirectory;
+
+    public ExtensionsTestInfrastructureDiscoverer(ExtensionManagerAdapter extensionManager, File generatedResourcesDirectory)
+    {
+        try
+        {
+            this.extensionManager = extensionManager;
+            this.generatedResourcesDirectory = generatedResourcesDirectory;
+            createManifestFileIfNecessary(generatedResourcesDirectory);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException("Error creating discoverer", e);
+        }
+    }
+
+    public ExtensionsTestInfrastructureDiscoverer(File generatedResourcesDirectory)
+    {
+        this(new DefaultExtensionManager(), generatedResourcesDirectory);
+        ((DefaultExtensionManager) extensionManager).setMuleContext(new DefaultMuleContext()
+        {
+            @Override
+            public MuleRegistry getRegistry()
+            {
+                return new MuleRegistryHelper(new DefaultRegistryBroker(this), this);
+            }
+        });
+        try
+        {
+            ((DefaultExtensionManager) extensionManager).initialise();
+        }
+        catch (InitialisationException e)
+        {
+            throw new RuntimeException("Error while initialising the extension manager", e);
+        }
+    }
+
+    /**
+     * @param describers if empty it will use annotatedClasses param to build the describers
+     * @param annotatedClasses used to build the describers
+     * @return a {@link List} of the resources generated for the given describers or annotated classes
+     */
+    public List<GeneratedResource> discoverExtensions(Describer[] describers, Class<?>[] annotatedClasses)
+    {
+        if (ArrayUtils.isEmpty(describers))
+        {
+            if (!ArrayUtils.isEmpty(annotatedClasses))
+            {
+                describers = new Describer[annotatedClasses.length];
+                int i = 0;
+                for (Class<?> annotatedClass : annotatedClasses)
+                {
+                    describers[i++] = new AnnotationsBasedDescriber(annotatedClass, new StaticVersionResolver(getProductVersion()));
+                }
+            }
+        }
+
+        if (ArrayUtils.isEmpty(describers))
```

Comment:
Changed to validate the case when no describers are passed as param and annotations classes param are used to build the describers, if no describer is created it will throw an IllegalStateException.